### PR TITLE
feat: add branch tags, remove version tags for non-releases

### PR DESCRIPTION
**Pull request by aaron-steinfeld**

### CODE REVIEW

File: hypertrace-gradle-docker-plugin/src/main/java/org/hypertrace/gradle/docker/DockerPlugin.java
Comment (adriancole) [inline hypertrace-gradle-docker-plugin/src/main/java/org/hypertrace/gradle/docker/DockerPlugin.java:107]: if I understand  correctly this "local" will be the image tag when run from normal checkout? while a bit of a bikeshed I prefer "test" if there isnt a strong opinion.

Comment (adriancole) [reply hypertrace-gradle-docker-plugin/src/main/java/org/hypertrace/gradle/docker/DockerPlugin.java:107]: maybe add a comment that this will result in "master" tag on normal CI and "local" ( or "test" ) otherwise as I think that is the intent but might not be obvious.

**Review by adriancole — COMMENTED** [review]
I think I get it.

File: hypertrace-gradle-docker-plugin/src/main/java/org/hypertrace/gradle/docker/DockerPlugin.java
Comment (aaron-steinfeld) [reply hypertrace-gradle-docker-plugin/src/main/java/org/hypertrace/gradle/docker/DockerPlugin.java:107]: Correct - I was going to use the branch name, but it didn't seem worth introducing a git dependency/complexity for the dev build case. No strong opinions on the specific label, so I'll switch it to "test" and add a comment.

File: hypertrace-gradle-docker-plugin/src/main/java/org/hypertrace/gradle/docker/DockerPlugin.java
Comment (adriancole) [inline hypertrace-gradle-docker-plugin/src/main/java/org/hypertrace/gradle/docker/DockerPlugin.java:107]: sorry to nit, but this rephrases the optional, which is literally correct, but doesn't add value I meant. if not saying "for example, "master"" It is likely better to not have a comment at all.

My reason for asking is as I don't know our CI setup, I don't know if we routinely build branches besides master. If the net effect of this change is that we end up with "master" tag, either here or in the CHANGELOG instead is probably more helpful from a product POV even if I agree what you are saying is technically correct.

**Review by adriancole — APPROVED** [review]
Looks good and feel free to address the comment as you see fit (or skip it as it isn't hugely important)

File: hypertrace-gradle-docker-plugin/src/main/java/org/hypertrace/gradle/docker/DockerPlugin.java
Comment (aaron-steinfeld) [reply hypertrace-gradle-docker-plugin/src/main/java/org/hypertrace/gradle/docker/DockerPlugin.java:107]: I actually had the example in here and removed it because of the technicality. First, the docker tag is actually used both in pull requests and a master build today, so even pull requests would _technically_ use this (I think it's fine not to explain that case). The part I was hesitating more on is that I don't believe the default branch is consistent across all repositories. I know we've been starting to move repository default branches from `master` to `main`. As that happens, we'll end up in a state where the tags are different across different containers - not ideal. So I tried to account for that here - maybe it would be better to just always use some known label rather than the branch name then?

Comment (adriancole) [reply hypertrace-gradle-docker-plugin/src/main/java/org/hypertrace/gradle/docker/DockerPlugin.java:107]: I had no idea about the move from master to main.. such a thing would need to be sorted (including if main is a good name, and here isn't a good place to discuss that).

In any case, I think there is a flex to things like this which is why some policy around tagging and which branches to build or tag end up in config, as filtering before executing a build can be more explicit and sometimes easier to manage.

**Comment by aaron-steinfeld** [comment]
Holding this one until we decide what name to apply to non-release images.

**Comment by adriancole** [comment]
@jcchavezs @pavolloffay I think you had some opinions on this so pinging you, but in any case we need to unlock this topic soon as the information won't change anymore.

"latest" tag is always last release, and the last release corresponds to a git tag of the same name (ex 2.1.3)
Since the mainline branch name is "main" I think the most intuitive way to represent latest commit is to use the tag "main"

This is different than -SNAPSHOT as remember there are non-java repos out there. Main thing (pun intended) is we need to for a while at least suggest using last commit on docker-compose because there is a lot of work changing from @JBAhire @kotharironak @ravisingal particularly. Once things stabilize (maybe a month from now) we can go back to "latest" (default tag in docker)

Another option is to switch nothing until things are stable, recognizing that our release history will pile up with a lot of spurious versions due to the former release policy which seems to be a variant of [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/). I don't think this is a good idea as it increases the cleanup tasks we already need to do cc @dnielsen 

*ACTION TO TAKE* if you think "main" is a fine tag in docker for last commit on mainline (formerly known as master), add thumbsup to this comment. 

Screen Shot 2020-08-24 at 9 38 10 AM

If you don't care, don't do anything. If you don't like this, use thumbsdown, but do say why in a comment below with a counter-proposal others can vote on similarly.

**Comment by pavolloffay** [comment]
I agree with @adriancole. 

It's generally useful to have the latest changes incorporated in a docker image. We also used to have a separate docker repository (e.g. `-dev`) with SHA tag per commit in the master/main branch.

**Comment by jcchavezs** [comment]
> It's generally useful to have the latest changes incorporated in a docker image. We also used to have a separate docker repository (e.g. -dev) with SHA tag per commit in the master/main branch.

In the past what worked for me was to have one image being published for each successful build on a PR (then you can try early in case you wan't to try something in an env before merge). One image per commit seems a bit too much to me but we could also try. in any case I totally agree on having a separate repository for that like `-dev` or `-canary`.

**Comment by aaron-steinfeld** [comment]
I believe the behavior in this PR is aligned with https://github.com/hypertrace/hypertrace-gradle-docker-plugins/pull/14#issuecomment-678861121 , so going to go ahead with it. If that's not the case, let me know and will revisit.

**Comment by adriancole** [comment]
thanks for the help!

**Pull request merged**
